src/Raven.Server/Documents/PeriodicBackup/DirectUpload/DirectFileUploader.cs

```diff
@@ -0,0 +1,199 @@
+                break;
+
+            default:
+                throw new ArgumentOutOfRangeException($"Missing implementation for direct upload destination '{_destination}'");
+        }
+    }
+
+    public IDictionary<string, string> GetObjectMetadata(string folderName, string objKeyName)
+    {
+        switch (_destination)
+        {
+            case BackupConfiguration.BackupDestination.AmazonS3:
+                return GetObjectMetadataFromS3(_settings.S3Settings, folderName, objKeyName);
+
+            case BackupConfiguration.BackupDestination.Azure:
+                return GetObjectMetadataFromAzure(_settings.AzureSettings, folderName, objKeyName);
+
+            default:
+                throw new ArgumentOutOfRangeException($"Missing implementation for direct upload destination '{_destination}'");
+        }
+    }
+
+    private string ThreadName(string name)
+    {
+        return $"Delete retired attachment of database '{_settings.DatabaseName}' from {name} (task: '{_settings.TaskName}')";
+    }
+
+    public void CreateUploadTask(DocumentDatabase database, Stream attachmentStream, string objKeyName, CancellationToken token)
+    {
+        var targetName = _destination == BackupConfiguration.BackupDestination.AmazonS3 ? S3Name : AzureName;
+        var threadName = $"Upload retired attachment '{objKeyName}' of database '{_settings.DatabaseName}' from {targetName} (task: '{_settings.TaskName}')";
+        var thread = PoolOfThreads.GlobalRavenThreadPool.LongRunning(_ =>
+        {
+            try
+            {
+                ThreadHelper.TrySetThreadPriority(ThreadPriority.BelowNormal, threadName, _logger);
+                Sparrow.Utils.NativeMemory.EnsureRegistered();
+
+                AddInfo($"Starting the upload of retired attachment '{objKeyName}'.");
+
+                using (attachmentStream)
+                using (var stream = StreamForBackupDestination(database, string.Empty, objKeyName))
+                    attachmentStream.CopyTo(stream);
+
+            }
+            catch (Exception e)
+            {
+                var extracted = e.ExtractSingleInnerException();
+                var error = $"Failed the {threadName}.";
+                Exception exception = null;
+                if (extracted is OperationCanceledException)
+                {
+                    // shutting down or HttpClient timeout
+                    exception = TaskCancelToken.Token.IsCancellationRequested ? extracted : new TimeoutException(error, e);
+                }
+
+                _exceptions.Add(exception ?? new InvalidOperationException(error, e));
+            }
+        }, null, ThreadNames.ForUploadRetiredAttachment(threadName, _settings.DatabaseName, targetName, _settings.TaskName));
+
+        _threads.Add(thread);
+    }
+
+    public bool TryCleanFinishedThreads(Stopwatch sp, OperationCancelToken token)
```

Comment:
You are trying to handle max concurrent uploads, I guess. But that is not the way to do that.
You need to have tasks, not threads. 
And simply use a semaphore or the like.

Reply:
in case I will have task it will be much simplier to manage, will handle it here :
https://issues.hibernatingrhinos.com/issue/RavenDB-24606/Retire-attachments-Server-Side-Use-Task-instead-of-Thread-in-DirectFileUploader.cs-Phase-1